Reduce potential for query conflicts

barryhughes opened the pull request:
Hey all,

We've hit some further conflicts when The Events Calendar and Sensei run side by side. The work already done in #1438 definitely improved things, but the same core problem still occurs if (for example) The Events Calendar widgets are in place beside a static page.

This change resolves that problem and so far as I can establish keeps things functioning as expected in the `allow_course_archive_on_front_page()` method.

Summary:

* Only hook `allow_course_archive_on_front_page()` when required (ie, if a static front page is set)
* Avoid unhooking/rehooking self from within the same action, to avoid issues re this [core WP bug](https://core.trac.wordpress.org/ticket/17817)
* Shuffles the logic a little to further avoid running an extra query until it's really necessary

#1438

georgestephanis commented on includes/class-sensei-course.php:
```diff
@@ -100,8 +100,9 @@ public function __construct () {
 		// Allow course archive to be setup as the home page
-		add_action( 'pre_get_posts', array( $this, 'allow_course_archive_on_front_page' ) );
-
+		if ( (int) get_option( 'page_on_front' ) > 0 ) {
```
I worry about this -- Core's test is always simply `if ( get_option( 'page_on_front' ) )` -- 

https://github.com/WordPress/WordPress/blob/9193013158d02ecd51ab8aa52d594aa789ac9836/wp-includes/rewrite.php#L504
https://github.com/WordPress/WordPress/blob/25e66e4f1e5ab91235c2de2fe7d218d0a2f2997d/wp-admin/options-reading.php#L65

With this structured the way it is, if the option got set to something like `'five'` or an unexpected value that can't be parsed to an integer but evaluates to true, we could have a situation where two bits of code are parsing the same value differently.

I'd sooner just use `if ( get_option( 'page_on_front' ) )` barring a really good counter-argument.

barryhughes replied:
Looking at the _current_ code in master, we have this test:

`$possible_home_page->ID != get_option( 'page_on_front' )`

So, essentially, if `get_option( 'page_on_front' )` doesn't contain something that can be a post ID then for our purposes here we aren't interested. Core does something else, but then it's not testing for the same thing. 

That said, I don't mind revising this to the simpler test you suggested if that's what you prefer :+1:

barryhughes replied:
...it also basically carries over this existing piece of logic:

`0 == intval( get_option( 'page_on_front' ) )`

It just applies the test earlier on (and avoids adding a callback altogether if it isn't satisfied).

dbtlr replied:
Giving that `page_on_front` _should_ always be a post id in order to be valid, this makes sense to me.

georgestephanis commented on includes/class-sensei-course.php:
```diff
@@ -3068,40 +3069,40 @@ public static function is_prerequisite_complete( $course_id ){
-			return; // This only needs to run on the home page. See https://github.com/Automattic/sensei/issues/1438
+		// Bail if it's clear we're not looking at a static front page or if the $running flag is
+		// set @see https://github.com/Automattic/sensei/issues/1438
+		if ( $running || ! $query->is_main_query() || ! is_page() || is_admin() ) {
```
Is this the most efficient order of operations for the tests?  I'd sooner have the cheapest, most likely to short-circuit tests put first to optimize.

barryhughes replied:
So keeping `$running` in its current position seems sensible, since it doesn't entail a function call and is a pre-requisite to everything else. I also like the positioning of `! $query->is_main_query()` because for all but one query in any given request we'll short-circuit here.

It might make sense to remove `is_admin()` from this method altogether and instead unhook this callback completely when `admin_init` fires, though, thus sparing us these tests altogether during admin requests (or, we could use a different action - `init` or something else - as a stepping stone: ie if it's not a frontend request, don't hook up this action).

What are your thoughts on that? My own view is that could be done as part of a subsequent change distinct from this one (because the fact that this callback is being hooked up during admin requests is not new), but I'd love your feedback.

georgestephanis commented on includes/class-sensei-course.php:
```diff
@@ -3068,40 +3069,40 @@ public static function is_prerequisite_complete( $course_id ){
-		if ( $not_a_valid_static_front_page ) {
-			// add action again to check subsequent calls
-			add_action( 'pre_get_posts', array( $this, 'allow_course_archive_on_front_page' ) );
+		$running = false;
```
If this isn't going to run again for subsequent queries (as per line 3127) and the point of `$running` is to prevent it from recursing into itself, what's the point of toggling it back to false again?  Why not just leave it so the function doesn't get called again?

Or is it in a way that it can fail multiple times after this point, but only complete once?

barryhughes replied:
Great question!

Well, the original reason was it can fail multiple times but complete only once. However, the fact is I was blindly copying the same strategy used by the original code - on further reflection I'm pretty sure we can remove this:

* If we get to this point, then we know we're looking at the main query
* We don't care about subsequent queries

So you're right, there's no reason to set `$running` back to false and we can also unhook the callback at this point (there's zero value in it running for queries _after_ the main query, that I can think of).

georgestephanis commented:
Added some thoughts on a couple sections, haven't tested it personally, but it looks reasonable.  If we can get the early conditional straightened out and it's tested and verified, should be mergeable.

:+1:

barryhughes commented:
Thanks for the feedback @georgestephanis - let me know any further thoughts you have and I will in any case follow up with some adjustments early-ish next week.

hansschuijff commented:
I'm a user of TEC in combination with WooThemes Sensei and been running into this problem for some months now, not in the single-pages but in the event list widget. Tired of waiting I dove into it this weekend and noticed that the allow_course_archive_on_front_page() method stops TEC's pre_get_post() action/method to run after it's execution. 

Probably that is caused by the WP bug mentioned by Barry here, but for me as a sensei customer it is troublesome that a know issue like this isn't yet fixed by some workaround after months. For now, knowing the cause, it is workable to just stop the allow_course_archive_on_front_page() method from blocking or have the tec pre_get_posts () method execute before that, but it would be nice if waiting on the WP core bug to be fixed woothemes implements another way to prevent an endless loop in the allow_course_archive_on_front_page() method.

barryhughes commented:
Updated and simplified :+1: 

> If we can get the early conditional straightened out ... should be mergeable

I'm taking this to be a reference to `(int) get_option( 'page_on_front' ) > 0` (in the constructor): please note that per [my earlier note](https://github.com/Automattic/sensei/pull/1502#discussion_r71959204) I decided not to change this, as it is essentially the same logic we had before this change, just moved to an earlier point in time.

dbtlr commented:
Thanks @barryhughes, I agree with your logic and am going to mark this as ready to merge.

barryhughes commented:
Thanks @dbtlr (+everyone else of course) for your help on this one :+1:
